### ISSUE TITLE
Handle Rate limiting from August APIs

### ISSUE THREAD

**Issue by gopi-ar** [homebridge-plugins/homebridge-august#55]
### Problem

With 3 locks, I got rate limited (429 error) within minutes of setup (was trying to setup 2 duplicate child bridges with the same account so 6 locks in total). I also have to restart the bridge if discovery fails at startup. Had to wait nearly an hour for August to unblock my IP to try again. 

If August blocks an IP, they don't allow even the first validation SMS to go out..

### Solution

Perhaps we can disable any unnecessary API calls, or delay status updates? Present homebridge-august settings are at 360 seconds refresh interval.

### Alternatives

- Retry calls after a few minutes
- Handle these errors more gracefully with a notification on HomeBridge / ios?

### Additional context

I get these errors when the IP is ratelimited. 
FetchError: Status 429 (undefined): undefined
FetchError: Status 429 (undefined): undefined
[13/04/2023, 12:30:56] [August] Discover Devices: TypeError: Cannot convert undefined or null to object

**Comment by donavanbecker** [comment]
@gopi-ar Are you able to submit a PR? I was thinking of removing the refresh or setting it to  refresh once a day

**Comment by gopi-ar** [comment]
Ah, not enough JS experience I'm afraid. If anyone ever gets the time, here are some ideas:
https://github.com/home-assistant/core/pull/31558

**Comment by donavanbecker** [comment]
@gopi-ar, I have prevented the default RefreshRate from being lower the 1800 (5 mins) that should fix this.

**Comment by jlg89** [comment]
I have seven locks, and I get rate-limited during initial status query. Looking at the debug log, it appears to successfully pull status for six of the locks, and then gets rate-limited (Status 429) when it tries to pull the seventh.

**Comment by jlg89** [comment]
This refresh rate issue is obviously caused by an August server-side rate-limiting configuration, so all we can do is try to guess "how much is too much" and stay under that threshold. Which is stupid, but it is what it is.

First thing, anyone who owns August locks should email support@august.com and complain about the 429 "too many requests" errors, let them know that they're alienating a chunk of their customer base. Tell them that we need better third-party interoperability because the August app doesn't support the kinds of automations we require.

On the development side, trying to stay under the 429 radar...does the plugin take into consideration the total number of defined devices that are being refreshed at any given time? I know we can set RefreshRate to different intervals, but at some point they'll all coincide, and if too many devices get queued up together it will set off the 429s. I don't know the best answer for this, other than maybe limit the total number of devices per poll to, say, four? That should cover the vast majority of use cases, and folks like me, who have more than four locks, will just have to go a bit longer between refreshes.

Maybe track a timeSinceLastRefresh for each device, and whenever a refresh happens, just do the four devices with the largest timeSinceLastRefresh values?

Could even make a maxRefreshCount so the user could define how many devices to refresh at once, since who knows what August will do tomorrow.

**Comment by jlg89** [comment]
Another idea: the log currently shows "FetchError: Status 429 (undefined): undefined" when August rate-limits a connection. Maybe make it read something like, "FetchError: Status 429 (Too Many Requests): The August server has rate-limited the connection. Email support@august.com to inform them that they are breaking things."

**Comment by donavanbecker** [comment]
@jlg89 If you want to submit a PR, I will approve.

**Comment by jlg89** [comment]
> @jlg89 If you want to submit a PR, I will approve.

I do various kinds of coding, and I would love to be able to submit a PR...but I haven't dug into this stuff, and reading through the code for this plugin I haven't even found where I'd start implementing what I've proposed, much less the exact syntax. In this case I'm the annoying guy who comes up with ideas, and every now and then one of them might actually be useful. 

There are a number of August plugins for homebridge, but this one is hands-down the best, and I have to think that the reason for that is that you did the coding.

If it would help, I can certainly contribute some funding toward this, knowing that financial compensation, though appreciated, is generally not an incentive. Time is the thing that's at a premium for everyone, there's just not enough of it. I'm not against learning how to fish, but if you don't have time to code it yourself, you sure don't have time to teach me how to code it. :P

**Comment by jlg89** [comment]
FWIW, I'm in the process of running this up the support flagpole at August. I finally managed to get past frontline support and escalated to the development team. Fingers crossed that they'll get back to me. I realize I may be an edge case given the number of locks at this facility, but I hope they will see the value in supporting an edge case.

**Comment by jlg89** [comment]
Update: Since the august-api module isn't officially supported by August, but SmartThings integration is, I decided to try that route. Turns out the SmartThings integration uses the same API (login & get a code), and my seven-lock setup breaks SmartThings integration too...but now August has to help, because they officially support SmartThings. I have both August and SmartThings tech support chasing down this issue, and I've let the SmartThings folks know that the problem is almost certainly with August's servers.

I'm hopeful that this will force August to relax their rate-limiting across the board, which would fix the august-api issue as well as the SmartThings issue. But they could also whitelist the SmartThings servers and/or apply different rate-limiting parameters for SmartThings, which wouldn't help august-api. Keep your fingers crossed.

**Comment by jlg89** [comment]
It took the SmartThings system several days to fully connect to all seven locks, but they seem to be working at this point. It was an interesting process. At first two locks worked, the rest showed "connected" in SmartThings but had no status. A couple days later, five of the seven were working. Finally, today, all seven were working.

I haven't heard boo from SmartThings support, and August support just keeps telling me that it's a SmartThings issue -- but I've been giving them some grief about it. I'm hoping to be able to have some dialogue with the SmartThings second-tier support folks, try to find out if they were in fact seeing 429 errors, maybe get some insight on how their system handles that...but I'm not holding my breath.

So, at the moment, if you have more than five August locks at a location, you will probably have to use the SmartThings integration, then use  https://github.com/iklein99/homebridge-smartthings (instead of homebridge-august) to pull them into homebridge. On the upside, it does have full support for everything, DoorSense, battery level, etc.

**Comment by jlg89** [comment]
Just an update. I went back and forth with August support and SmartThings support for weeks. 

What I got from August Support:

1. The problem is with your device.
2. No, the problem is with the SmartThings app.
3. No, the problem is with the SmartThings system.
4. Well, maybe the problem is with the August servers, but SmartThings support has to contact someone at August to figure it out.

SmartThings Support said:

"Our Tier 2 Support have sought assistance directly from our Tier 3 and based on their investigation, they found out on the backend cloud logs that August is returning a StateRefreshRequest failed response. The details on this error state August Home returned an empty response.

They are recommending to reach out to August Home to investigate why the state refresh response is failing."

When I forwarded that question to August Support, that's when they said that SmartThings Tier 2/3 should contact their August Support liaison directly. And that's where things stand at this point.
